python/context_selector/context_widget.py

```diff
@@ -0,0 +1,956 @@
+        self.ui.task_search.hide()
+
+        # setup the search toggle
+        self.ui.link_search_btn.toggled.connect(self._on_link_search_toggled)
+        self.ui.link_search.hide()
+
+        # set up the task manager to the task search widget
+        self.ui.task_search.set_placeholder_text("Search for Tasks...")
+        self.ui.task_search.set_bg_task_manager(task_manager)
+        self.ui.task_search.entity_activated.connect(
+            self._on_entity_activated)
+
+        # save as above but for the link search widget
+        self.ui.link_search.set_placeholder_text("Search for entity link...")
+        self.ui.link_search.set_bg_task_manager(task_manager)
+        self.ui.link_search.entity_activated.connect(
+            self._on_entity_activated
+        )
+
+        # set up event filters for the task/link display labels so that when
+        # clicked, they go directly to an edit state
+        self.ui.task_display.installEventFilter(self)
+        self.ui.link_display.installEventFilter(self)
+
+        # setup event filters for the task/link search inputs so that when
+        # certain keys are pressed, the widget can react to it properly
+        self.ui.task_search.installEventFilter(self)
+        self.ui.link_search.installEventFilter(self)
+
+        # we need to limit the search completer to entity types that are valid
+        # for ``PublishedFile.entity`` field links. To do this, query the
+        # shotgun schema to get a list of these entity types. We use the current
+        # project schema if we're in a project. We do this as a background query
+        # via the supplied task manager.
+
+        # connect to the task manager signals so that we can get the results
+        task_manager.task_completed.connect(self._on_task_completed)
+        task_manager.task_failed.connect(self._on_task_failed)
+
+        # Query Shotgun for valid entity types for PublishedFile.entity field
+        self._schema_query_id = task_manager.add_task(
+            _query_publishedfile_entity_schema)
+
+        # query all my assigned tasks in a bg task
+        self._my_tasks_query_id = task_manager.add_task(_query_my_tasks)
+
+        # get recent contexts from user settings
+        self._get_recent_contexts()
+
+    @property
+    def context_label(self):
+        """
+        The label for the context widget.
+        """
+        return self.ui.label
+
+    def enable_editing(self, enabled, message=None):
+        """
+        Show/hide the input widgets and display a message in the context label.
+        """
+
+        if enabled:
+            self.ui.edit_widget.show()
+        else:
+            self.ui.edit_widget.hide()
+
+        self.context_label.setText(message or "")
+
+    def _add_to_recents(self, context):
+        """
+        Adds the supplied context as an action in the list of recents context
+        actions
+        """
+
+        # don't add a "project" context to recents. we shouldn't encourage it
```

Comment:
publisher specific?

Reply:
I think it's fine to keep it for now. i agree with the comment :)